linalg/tensor.hpp

```diff
@@ -0,0 +1,2051 @@
+ * subtracting the mean stress (average of main diagonal elements)
+ * from each element on the main diagonal
+ */
+template <typename T, int n> MFEM_HOST_DEVICE
+tensor<T, n, n> dev(const tensor<T, n, n>& A)
+{
+   auto devA = A;
+   auto trA  = tr(A);
+   for (int i = 0; i < n; i++)
+   {
+      devA[i][i] -= trA / n;
+   }
+   return devA;
+}
+
+/**
+ * @brief Obtains the identity matrix of the specified dimension
+ * @return I_dim
+ */
+template <int dim>
+MFEM_HOST_DEVICE tensor<double, dim, dim> Identity()
+{
+   tensor<double, dim, dim> I{};
+   for (int i = 0; i < dim; i++)
+   {
+      for (int j = 0; j < dim; j++)
+      {
+         I[i][j] = (i == j);
+      }
+   }
+   return I;
+}
+
+/**
+ * @brief Returns the transpose of the matrix
+ * @param[in] A The matrix to obtain the transpose of
+ */
+template <typename T, int m, int n> MFEM_HOST_DEVICE
+tensor<T, n, m> transpose(const tensor<T, m, n>& A)
+{
+   tensor<T, n, m> AT{};
+   for (int i = 0; i < n; i++)
+   {
+      for (int j = 0; j < m; j++)
+      {
+         AT[i][j] = A[j][i];
+      }
+   }
+   return AT;
+}
+
+/**
+ * @brief Returns the determinant of a matrix
+ * @param[in] A The matrix to obtain the determinant of
+ */
+template <typename T> MFEM_HOST_DEVICE
+T det(const tensor<T, 2, 2>& A)
+{
+   return A[0][0] * A[1][1] - A[0][1] * A[1][0];
+}
+/// @overload
+template <typename T> MFEM_HOST_DEVICE
+T det(const tensor<T, 3, 3>& A)
+{
+   return A[0][0] * A[1][1] * A[2][2] + A[0][1] * A[1][2] * A[2][0] + A[0][2] *
+          A[1][0] * A[2][1] -
+          A[0][0] * A[1][2] * A[2][1] - A[0][1] * A[1][0] * A[2][2] - A[0][2] * A[1][1] *
+          A[2][0];
+}
+
+/**
+ * @brief Return whether a square rank 2 tensor is symmetric
+ *
+ * @tparam n The height of the tensor
+ * @param A The square rank 2 tensor
+ * @param tolerance The tolerance to check for symmetry
+ * @return Whether the square rank 2 tensor (matrix) is symmetric
+ */
+template <int n> MFEM_HOST_DEVICE
+bool is_symmetric(tensor<double, n, n> A, double tolerance = 1.0e-8)
+{
+   for (int i = 0; i < n; ++i)
+   {
+      for (int j = i + 1; j < n; ++j)
+      {
+         if (std::abs(A(i, j) - A(j, i)) > tolerance)
+         {
+            return false;
+         }
+      }
+   }
+   return true;
+}
+
+/**
+ * @brief Return whether a matrix is symmetric and positive definite
+ * This check uses Sylvester's criterion, checking that each upper left subtensor has a
+ * determinant greater than zero.
+ *
+ * @param A The matrix to test for positive definiteness
+ * @return Whether the matrix is positive definite
+ */
+inline MFEM_HOST_DEVICE
+bool is_symmetric_and_positive_definite(tensor<double, 2, 2> A)
+{
+   if (!is_symmetric(A))
+   {
+      return false;
+   }
+   if (A(0, 0) < 0.0)
+   {
+      return false;
+   }
+   if (det(A) < 0.0)
+   {
+      return false;
+   }
+   return true;
+}
+/// @overload
+inline MFEM_HOST_DEVICE
+bool is_symmetric_and_positive_definite(tensor<double, 3, 3> A)
```

Comment:
I don't see a good reason to have these `is_symmetric*` functions here. I recommend removing them.